src/cdk/scrolling/virtual-for-of.ts

```diff
@@ -0,0 +1,310 @@
+        new StaticArrayDataSource<T>(Array.prototype.slice.call(value));
+    this._dataSourceChanges.next(ds);
+  }
+  _cdkVirtualForOf: NgIterable<T> | DataSource<T>;
+
+  /**
+   * The `TrackByFunction` to use for tracking changes. The `TrackByFunction` takes the index and
+   * the item and produces a value to be used as the item's identity when tracking changes.
+   */
+  @Input()
+  get cdkVirtualForTrackBy(): TrackByFunction<T> {
+    return this._cdkVirtualForTrackBy;
+  }
+  set cdkVirtualForTrackBy(fn: TrackByFunction<T>) {
+    this._needsUpdate = true;
+    this._cdkVirtualForTrackBy =
+        (index, item) => fn(index + (this._renderedRange ? this._renderedRange.start : 0), item);
+  }
+  private _cdkVirtualForTrackBy: TrackByFunction<T>;
+
+  /** The template used to stamp out new elements. */
+  @Input()
+  set cdkVirtualForTemplate(value: TemplateRef<CdkVirtualForOfContext<T>>) {
+    if (value) {
+      this._needsUpdate = true;
+      this._template = value;
+    }
+  }
+
+  /** Emits whenever the data in the current DataSource changes. */
+  dataStream: Observable<T[]> = this._dataSourceChanges
+      .pipe(
+          // Start off with null `DataSource`.
+          startWith(null!),
+          // Bundle up the previous and current data sources so we can work with both.
+          pairwise(),
+          // Use `_changeDataSource` to disconnect from the previous data source and connect to the
+          // new one, passing back a stream of data changes which we run through `switchMap` to give
+          // us a data stream that emits the latest data from whatever the current `DataSource` is.
+          switchMap(([prev, cur]) => this._changeDataSource(prev, cur)),
+          // Replay the last emitted data when someone subscribes.
+          shareReplay(1));
+
+  private _differ: IterableDiffer<T> | null = null;
+
+  private _data: T[];
+
+  private _renderedItems: T[];
+
+  private _renderedRange: Range;
+
+  private _templateCache: EmbeddedViewRef<CdkVirtualForOfContext<T>>[] = [];
+
+  private _needsUpdate = false;
+
+  constructor(
+      private _viewContainerRef: ViewContainerRef,
+      private _template: TemplateRef<CdkVirtualForOfContext<T>>,
+      private _differs: IterableDiffers,
+      @Host() private _viewport: CdkVirtualScrollViewport) {
+    this.dataStream.subscribe(data => this._data = data);
+    this._viewport.renderedRangeStream.subscribe(range => this._onRenderedRangeChange(range));
+    this._viewport.connect(this);
+  }
+
+  /**
+   * Get the client rect for the given index.
+   * @param index The index of the data element whose client rect we want to measure.
+   * @return The combined client rect for all DOM elements rendered as part of the given index.
+   *     Or null if no DOM elements are rendered for the given index.
+   * @throws If the given index is not in the rendered range.
+   */
+  measureClientRect(index: number): ClientRect | null {
```

Comment:
You might want to add a window resize function to update dims on window resizes.

Reply:
Responding to resize is a feature I plan to add, but not in this PR